Test fails due to stack traces

Hi, tests are failing again since go 1.11.

Maybe @gabrielf could come up again with a solution for this? :P

Thanks a lot.

```
$ go version
go version go1.11.5 linux/amd64

$ go test -vet=off -v -p 4 github.com/go-errors/errors
=== RUN   TestStackFormat
--- FAIL: TestStackFormat (0.00s)
    error_test.go:33: Stack trace does not contain source line: 'a: b(5)'
    error_test.go:34: /build/golang-github-go-errors-errors-1.0.1/obj-x86_64-linux-gnu/src/github.com/go-errors/errors/error_test.go:21 (0x4f5f35)
        	TestStackFormat.func1: e, expected := Errorf("hi"), callers()
        /usr/lib/go-1.11/src/runtime/asm_amd64.s:522 (0x455b3b)
        	call32: CALLFN(·call32, 32)
        /usr/lib/go-1.11/src/runtime/panic.go:513 (0x429df9)
        	gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
        /build/golang-github-go-errors-errors-1.0.1/obj-x86_64-linux-gnu/src/github.com/go-errors/errors/error_test.go:245 (0x4f38ff)
        	TestStackFormat: panic('a')
        /usr/lib/go-1.11/src/testing/testing.go:827 (0x4b573f)
        	tRunner: fn(t)
        /usr/lib/go-1.11/src/runtime/asm_amd64.s:1333 (0x4576d1)
        	goexit: BYTE	$0x90	// NOP
```

Originally reported in Debian [[1]] by Adrian Bunk.

[1]: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=918619#5

> > For some reasons, this test seems to pass with Go 1.12, but continues to fail with Go 1.11.
> 
> Great news. Thanks for the fix in Debian.

You're very welcome.  I needed the fix myself anyway.  :-)

The stack trace test seems very flaky; or, rather, Go sometimes generate unpredictable stack trace?  I am seeing different stack trace outputs with gccgo-8 and gccgo-9, for example, on amd64 vs mipsel.

* gccgo-8 on amd64: FAIL.
* gccgo-9 on amd64: PASS.  Yay!
* gccgo-8 on amd64: TestStackFormat passes, but TestNew fails:

```
--- FAIL: TestNew (0.00s)
	error_test.go:81: Stack didn't match
	error_test.go:82: First entry PC diff to large (36)
		Actual stack trace:
		/home/foka/golang-github-go-errors-errors-1.0.1/error_test.go:78 (0x5559d307)
			__home_foka_golang_github_go_errors_errors_1_0_1.TestNew
		../../../src/libgo/go/testing/testing.go:777 (0x77924fbb)
			testing.tRunner
		../../../src/libgo/go/runtime/proc.go:1167 (0x778a2293)
			runtime.kickoff
		
		Expected stack trace:
		/home/foka/golang-github-go-errors-errors-1.0.1/error_test.go:78 (0x5559d32b)
			__home_foka_golang_github_go_errors_errors_1_0_1.TestNew
		../../../src/libgo/go/testing/testing.go:777 (0x77924fbb)
			testing.tRunner
		../../../src/libgo/go/runtime/proc.go:1167 (0x778a2293)
			runtime.kickoff
```

* gccgo-9 on mipsel:

```
# _/home/foka/golang-github-go-errors-errors-1.0.1 [_/home/foka/golang-github-go-errors-errors-1.0.1.test]
go1: internal compiler error: in mangled_name, at go/gofrontend/names.cc:559
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-9/README.Bugs> for instructions.
FAIL	_/home/foka/golang-github-go-errors-errors-1.0.1 [build failed]
```

Oh well...

--- a/error_test.go
+++ b/error_test.go
@@ -10,6 +10,26 @@ import (
 	"testing"
 )
 
+func BenchmarkStackFormat(b *testing.B) {
+	b.ReportAllocs()
+
+	for i := 0; i < b.N; i++ {
+		func() {
+			defer func() {
+				err := recover()
+				if err != 'a' {
+					b.Fatal(err)
+				}
+
+				e := Errorf("hi")
+				_ = string(e.Stack())
+			}()
+
+			a()
+		}()
+	}
+}
+
 func TestStackFormat(t *testing.T) {
 
 	defer func() {

--- a/error_test.go
+++ b/error_test.go
@@ -247,7 +247,7 @@ func c() {
 
 // compareStacks will compare a stack created using the errors package (actual)
 // with a reference stack created with the callers function (expected). The
-// first entry is compared inexact since the actual and expected stacks cannot
+// first entry is not compared  since the actual and expected stacks cannot
 // be created at the exact same program counter position so the first entry
 // will always differ somewhat. Returns nil if the stacks are equal enough and
 // an error containing a detailed error message otherwise.
@@ -256,12 +256,7 @@ func compareStacks(actual, expected []uintptr) error {
 		return stackCompareError("Stacks does not have equal length", actual, expected)
 	}
 	for i, pc := range actual {
-		if i == 0 {
-			firstEntryDiff := (int)(expected[i]) - (int)(pc)
-			if firstEntryDiff < -27 || firstEntryDiff > 27 {
-				return stackCompareError(fmt.Sprintf("First entry PC diff to large (%d)", firstEntryDiff), actual, expected)
-			}
-		} else if pc != expected[i] {
+		if i != 0 && pc != expected[i] {
 			return stackCompareError(fmt.Sprintf("Stacks does not match entry %d (and maybe others)", i), actual, expected)
 		}
 	}

--- a/stackframe.go
+++ b/stackframe.go
@@ -1,9 +1,10 @@
 package errors
 
 import (
+	"bufio"
 	"bytes"
 	"fmt"
-	"io/ioutil"
+	"os"
 	"runtime"
 	"strings"
 )
@@ -62,18 +63,29 @@ func (frame *StackFrame) String() string {
 
 // SourceLine gets the line of code (from File and Line) of the original source if possible.
 func (frame *StackFrame) SourceLine() (string, error) {
-	data, err := ioutil.ReadFile(frame.File)
+	if frame.LineNumber <= 0 {
+		return "???", nil
+	}
 
+	file, err := os.Open(frame.File)
 	if err != nil {
 		return "", New(err)
 	}
+	defer file.Close()
 
-	lines := bytes.Split(data, []byte{'\n'})
-	if frame.LineNumber <= 0 || frame.LineNumber >= len(lines) {
-		return "???", nil
+	scanner := bufio.NewScanner(file)
+	currentLine := 1
+	for scanner.Scan() {
+		if currentLine == frame.LineNumber {
+			return string(bytes.Trim(scanner.Bytes(), " \t")), nil
+		}
+		currentLine++
 	}
-	// -1 because line-numbers are 1 based, but our array is 0 based
-	return string(bytes.Trim(lines[frame.LineNumber-1], " \t")), nil
+	if err := scanner.Err(); err != nil {
+		return "", New(err)
+	}
+
+	return "???", nil
 }
 
 func packageAndName(fn *runtime.Func) (string, string) {

